loop: add uv_loop_stats

jasnell opened the pull request:
This is a bit of speculative work in support of https://github.com/nodejs/node/pull/14680.

Adds `uv_loop_stats_t` and `uv_loop_stats()` APIs that provide high resolution timing for event loop phases. This approach introduces significantly less latency than using callback handles for the same purpose, and gives more precise results.

/cc @matthewloring

saghul commented on include/uv-unix.h:
```diff
@@ -220,6 +220,7 @@ typedef struct {
   uv__io_t signal_io_watcher;                                                 \
   uv_signal_t child_watcher;                                                  \
   int emfile_fd;                                                              \
+  uv_loop_stats_t stats;                                                      \
```
This breaks the ABI, so we can't land it on v1.x.

jasnell replied:
yeah, was a bit afraid of that.

saghul commented on docs/src/loop.rst:
```diff
@@ -218,3 +234,8 @@ API
        invalid. That function must be called again to determine the
        correct backend file descriptor.
+
+.. c:function:: uv_loop_stats_t* uv_loop_stats(uv_loop_t* loop)
```
this function should copy the stats instead of returning a pointer to an internal field.

jasnell replied:
+1 ... was wanting to avoid the copying if at all possible and went back and forth on this. Copying works.

saghul commented on include/uv.h:
```diff
@@ -1015,6 +1016,17 @@ struct uv_passwd_s {
   char* homedir;
 };
 
+struct uv_loop_stats_s {
```
what about async handles?

jasnell replied:
still thinking through those

saghul commented:
What is the desired information here? I see a bunch of _start fields, but no _stop. Also, currently they are pretty much all going to match the loop iteration count, because they run on every loop iteration.

Can you please elaborate on what are the desired metrics?

Last, this breaks the ABI, so it should be targeted at master.

jasnell commented:
> What is the desired information here?

Specifically the hr timestamps when each of the key phases kicks off. The original iteration of this included `_stop`s but they were so close to the subsequent `_start`s that they were useless. The goal is to be able to generate precise measurements of latency between phases and knowing precisely when those start is generally good enough.

Whether this is the right approach to take, I'm not yet sure. I'd much prefer to take an approach that doesn't break ABI but I'm still looking at it.

saghul commented:
> The original iteration of this included _stops but they were so close to the subsequent _starts that they were useless. The goal is to be able to generate precise measurements of latency between phases and knowing precisely when those start is generally good enough.

This makes assumptions about the order of phases, which is something I really don't want people to rely on. Also, about them being so close: this could be used as a mechanism to tell the user that too much time was spent in a callback (I can imagine some computations taking too long on a timer, or a sync fs operation,...).

jasnell commented:
Makes sense. Starts and stops for each phase then, without any assumption about ordering.

jasnell commented:
@saghul @mcollina ... PTAL

I've refactored this significantly from the first take and rebased the PR against master.

This enables two basic models for getting stats data: sync and async. The sync is pretty straight forward:

```c
  uv_stats_info_t info;
  uv_loop_stats(uv_default_loop(), &info);

  printf("loop enter: %llu\n", info.loop_enter);
  printf("loop exit: %llu\n", info.loop_exit);
```

The async uses `uv_loop_configure` per @saghul's suggestion:

```c
void on_stats(uv_stats_info_t* info) { /** ... **/ }

uv_stats_config_t config = { UV_LOOP_STATS_TICK, 0, on_stats };
uv_loop_configure(uv_default_loop(), UV_LOOP_STATS, &config);
uv_run(uv_default_loop(), UV_RUN_DEFAULT);
```

There are three scheduling options configured using the `uv_stats_config_t` struct:

* `UV_LOOP_STATS_TICK` - notify on every turn
* `UV_LOOP_STATS_COUNT` - notify on every Nth turn
* `UV_LOOP_STATS_TIME` - notify every Nth nanosecond

For now the start and end timestamps for each loop phase, the enter and exit timestamps for the loop, and the number of loop turns are being tracked. We can add aditional measures (such as the number of pending events) fairly easily.

mcollina commented:
I'm really 👍  with this, too bad it needs to wait for libuv 2.

jasnell commented:
I would *really* like to try to find a way of doing this that doesn't break ABI. If we can find a non-terrible and *efficient* way of having a separate loop->stats map that doesn't require adding information to `uv_loop_t`, then it's possible.

santigimeno commented:
There's this [`data` field](https://github.com/libuv/libuv/blob/v1.x/include/uv.h#L1489) that _might_ be used? Not sure if it's a good idea though

saghul commented:
> Not sure if it's a good idea though

That field is for users, not our own internal use.

pfreixes commented:
Hi @jasnell I was keen on implementing a feature to measure the loop load [1], @saghul told me that this feature might be built on top of your PR. I definitely believe that yes, it can be built on top of that feature. But I'm missing one statistic

To calculate a more precise load its needed use the maximum time spent polling IO between the timeout and the real time used by the OS. When the OS does not schedule the loop after the timeout time means that the was enough busy to don't give the  CPU at that moment, it becomes especially important when the CPU pressure is highly demanded by other processes that share the same CPU.

Having this PR, the code needed to calculate the load can be done in the userland. Therefore, I will focus on the userland code.

BTW I will like to see this change in the 1.X series.    

[1] https://groups.google.com/forum/#!topic/libuv/RVtgltLI_fE

pfreixes commented:
@jasnell I've managed to implement a userland code written in Cython that uses your feature, looks promising. Two comments.

The stats callback should provide as an argument the loop instance, this will give the chance to the user to retrieve the context data that is set by the user once the loop is initialized in the `data` attribute. 

About performance, using Cython the degradation experimented using the UV_LOOP_STATS_TICK is less than 3%. Having in mind that peak reached in my computer is almost 300K coroutines per second I believe that this is far far away that it would be needed in a Python environment. Therefore I can implement the load strategy on top of this feature.

PD: remember what I said about the timeout, if it might be saved as an attribute of the stats this will help a lot.

bnoordhuis commented:
This PR needs a rebase and overlaps to some extent with #1528.  The APIs can probably be unified a little.

jasnell commented:
Will definitely take a look when I'm back from node interactive. Would make a good activity for the flight home. 

Any particular concerns beyond reconciling with #1528?

pfreixes commented:
Yeps, any news about that @jasnell? I would like to see this MR moving ahead :) but having also a new metric with the timeout used during the IO poll, somehting that should be a trivial change.

jasnell commented:
Updating this now. Will work on it with a separate PR, however.

pfreixes commented:
@jasnell another thing that I forgot to mention

> The stats callback should provide as an argument the loop instance, this will give the chance to the user to retrieve the context data that is set by the user once the loop is initialized in the data attribute.

jasnell commented:
Replaced by https://github.com/libuv/libuv/pull/1764
